pkgs/build-support/node/fetch-yarn-deps/yarn-install-hook.sh

```diff
@@ -0,0 +1,82 @@
+
+    # `npm pack` writes to cache so temporarily override it
+    while IFS= read -r file; do
+        local dest="$packageOut/$(dirname "$file")"
+        mkdir -p "$dest"
+        cp "$file" "$dest"
+    done < <(@jq@ --raw-output '.[0].files | map(.path | select(. | startswith("node_modules/") | not)) | join("\n")' <<< "$(npm_config_cache="$HOME/.npm" npm pack --json --dry-run --loglevel=warn --no-foreground-scripts $npmPackFlags "${npmPackFlagsArray[@]}" $npmFlags "${npmFlagsArray[@]}")")
+
+    # Based on code from Python's buildPythonPackage wrap.sh script, for
+    # supporting both the case when makeWrapperArgs is an array and a
+    # IFS-separated string.
+    #
+    # TODO: remove the string branch when __structuredAttrs are used.
+    if [[ "${makeWrapperArgs+defined}" == "defined" && "$(declare -p makeWrapperArgs)" =~ ^'declare -a makeWrapperArgs=' ]]; then
+        local -a user_args=("${makeWrapperArgs[@]}")
+    else
+        local -a user_args="(${makeWrapperArgs:-})"
+    fi
+    while IFS=" " read -ra bin; do
+        mkdir -p "$out/bin"
+        makeWrapper @hostNode@ "$out/bin/${bin[0]}" --add-flags "$packageOut/${bin[1]}" "${user_args[@]}"
+    done < <(@jq@ --raw-output '(.bin | type) as $typ | if $typ == "string" then
+        .name + " " + .bin
+        elif $typ == "object" then .bin | to_entries | map(.key + " " + .value) | join("\n")
+        elif $typ == "null" then empty
+        else "invalid type " + $typ | halt_error end' "package.json")
+
+    while IFS= read -r man; do
+        installManPage "$packageOut/$man"
+    done < <(@jq@ --raw-output '(.man | type) as $typ | if $typ == "string" then .man
+        elif $typ == "list" then .man | join("\n")
+        elif $typ == "null" then empty
+        else "invalid type " + $typ | halt_error end' "package.json")
+
+    local -r nodeModulesPath="$packageOut/node_modules"
+
+    if [ ! -d "$nodeModulesPath" ]; then
+        if [ -z "${dontYarnPrune-}" ]; then
+            # Yarn has a 'prune' command, but it's only a stub that directs you to use install
+            if ! yarn install \
+                --frozen-lockfile \
+                --force \
+                --production=true \
+                --ignore-engines \
+                --ignore-platform \
+                --ignore-scripts \
+                --no-progress \
+                --non-interactive \
+                --offline
+            then
+              echo
+              echo
+              echo "ERROR: yarn prune step failed"
+              echo
+              echo 'If yarn tried to download additional dependencies above, try setting `dontYarnPrune = true`.'
+              echo
+
+              exit 1
+            fi
+        fi
+
+        find node_modules -maxdepth 1 -type d -empty -delete
+
+        cp -r node_modules "$nodeModulesPath"
+    fi
+
+    runHook postInstall
+
+    echo "Finished yarnInstallHook"
+}
+
+if [ -z "${dontYarnInstall-}" ] && [ -z "${installPhase-}" ]; then
```

Comment:
Or `dontYarnPrune`?

Reply:
Both, they do different things, and there's two flags on purpose.

Looking at the code might make things more confusing since `dontYarnPrune` skips running `yarn install` and `dontYarnInstall` skips running things which are not `yarn install`.

It might be better to think of the abstraction that's being created here:

- `yarnInstallHook` cleans up devDependencies, installs executables and man pages.
- Don't want to run any of that? `dontYarnInstall = true`
- Don't want to remove devDependencies, but otherwise run the normal installation? `dontYarnPrune = true`

Reply:
> Both, they do different things, and there's two flags on purpose.
> 
> Looking at the code might make things more confusing since `dontYarnPrune` skips running `yarn install` and `dontYarnInstall` skips running things which are not `yarn install`.

I'd suggest a different naming:

- Don't check `dontYarnInstall` here - where you set `installPhase`.
- Check `dontYarnInstall` only in the `yarnInstallHook` - where now you check `dontInstallPrune`.

If people will want from some reason to add the `yarnInstallHook` to their `nativeBuildInputs` and not use the hook directly (or not use it at all), they can always override the `installPhase`.

Reply:
I don't like this, I've personally never seen a hook that cannot be disabled like this.
Also other yarn hooks follow the same nomenclature dontYarn{Configure,Build}, I don't see how any name other than `dontYarnInstall` would make sense.

The name `dontYarnPrune` was chosen to follow `dontNpmPrune`.

I think we're talking about the same thing on two threads :smiley: